src/storage/storage.ts

```diff
@@ -0,0 +1,634 @@
+            authorities: 1,
+            value: 1,
+            address: addressForChange,
+            timelock: null,
+          });
+        }
+      }
+
+      // match melt
+      if (balance.melt > 0) {
+        if (!chooseInputs) {
+          // We cannot choose inputs, so we fail
+          throw new Error(`Insufficient authority in the given inputs for ${token}, missing ${balance.melt} more melt authority utxos.`);
+        }
+        // We have a surplus of this token on the outputs, so we need to find utxos to match
+        let foundAmount = 0;
+        // We use max_utxos to find at most `balance.melt` inputs
+        for await (const utxo of this.selectUtxos({ token, authorities: 1, max_utxos: balance.melt})) {
+          foundAmount += 1;
+          newInputs.push({
+            txId: utxo.txId,
+            index: utxo.index,
+            token: utxo.token,
+            address: utxo.address,
+            value: utxo.value,
+            authorities: utxo.authorities,
+          });
+        }
+        if (foundAmount < balance.melt) {
+          // XXX: Insufficient funds
+          throw new Error('Insufficient melt authority');
+        }
+        // Since we use max_utxos to stop before we select more the the
+        // balance requires there is no need to add an authority change
+      } else if (balance.melt < 0) {
+        // We have a surplus of this token on the inputs, so we need to add enough change outputs to match
+        for (let i = 0; i < Math.abs(balance.melt); i++) {
+          newOutputs.push({
+            type: getAddressType(addressForChange, this.config.getNetwork()),
+            token,
+            authorities: 2,
+            value: 2,
+            address: addressForChange,
+            timelock: null,
+          });
+        }
+      }
+    }
+
+    const max_inputs = this.version?.max_number_inputs || 255;
+    const max_outputs = this.version?.max_number_outputs || 255;
+    if (((tx.inputs.length + newInputs.length) > max_inputs)
+      || ((tx.outputs.length + newOutputs.length) > max_outputs)
+    ) {
+      // we have more inputs/outputs than what can be sent on the transaction
+      throw new Error('When over the maximum amount of inputs/outputs');
+    }
+
+    // for (const input of newInputs) {
+    //   tx.inputs.push(input);
+    // }
+    // for (const output of tx.outputs) {
+    //   tx.outputs.push(output);
+    // }
+
+    return {inputs: newInputs, outputs: newOutputs}
+  }
+
+  /**
+   * Mark an utxo as selected as input
+   *
+   * @param {IUtxoId} utxo The Data to identify the utxo
+   * @param {boolean} markAs Mark the utxo as this value
+   * @param {number|undefined} ttl Unmark the utxo after this amount os ms passed
+   *
+   * @async
+   * @returns {Promise<void>}
+   */
+  async utxoSelectAsInput(utxo: IUtxoId, markAs: boolean, ttl?: number): Promise<void> {
+    const tx = await this.getTx(utxo.txId);
+    if ((!tx) || (!tx.outputs[utxo.index])) {
+      return;
```

Comment:
That was the original implementation, but some other places would break.
For instance, the `sendTransaction` facade marks the utxos as  `selected_as_input`, but this facade also works with atomic swaps, meaning they will try to mark utxos not from our wallet.

I decided that failing silently for utxos not from our wallet and utxos already spent is acceptable since it was the same behavior as before.